src/backend/opencl/kernel_cache.hpp

```diff
@@ -0,0 +1,101 @@
+    /// essentially forwarded to enqueueKernelLaunch API call.
+    ///
+    /// \param[in] qArgs is an object of struct \ref cl::EnqueueArgs
+    /// \param[in] args is the placeholder for variadic arguments
+    ///
+    template<typename... Args>
+    void operator()(const cl::EnqueueArgs& qArgs, Args... args) {
+        auto launchOp = cl::KernelFunctor<Args...>(*ker);
+        launchOp(qArgs, std::forward<Args>(args)...);
+    }
+};
+
+///
+/// \brief Find/Create-Cache a Kernel that fits the given criteria
+///
+/// This function takes in two vectors of strings apart from the main Kernel
+/// name, match criteria, to find a suitable kernel in the Kernel cache. It
+/// builds and caches a new Kernel object if one isn't found in the cache.
+///
+/// The paramter \p key has to be the unique name for a given OpenCL kernel.
+/// The key has to be present in one of the entries of KernelMap defined in
+/// the header EnqueueArgs.hpp.
+///
+/// The parameter \p templateArgs is a list of stringified template arguments of
+/// the OpenCL kernel. These strings are used to generate the template
+/// instantiation expression of the OpenCL kernel during compilation stage. This
+/// string is used as key to kernel cache map. At some point in future, the idea
+/// is to use these instantiation strings to generate template instatiations in
+/// OpenCL online compiler.
+///
+/// The paramter \p compileOpts is a list of strings that lets you add
+/// definitions such as `-D<NAME>` or `-D<NAME>=<VALUE>` to the compiler. To
+/// enable easy stringification of variables into their definition equation,
+/// three helper macros are provided: TemplateArg, DefineKey and DefineValue.
+///
+/// Example Usage: transpose
+///
+/// \code
+/// static const std::string src(transpose_cuh, transpose_cuh_len);
+/// auto transpose = getKernel("cuda::transpose", src,
+///         {
+///           TemplateTypename<T>(),
+///           TemplateArg(conjugate),
+///           TemplateArg(is32multiple)
+///         },
+///         {
+///           DefineValue(THREADS_Y) // Results in a definition
+///                                  // "-D THREADS_Y=<Value of THREADS_Y>"
+///           DefineKeyValue(DIMY, threads_y)  // Results in a definition
+///                                            // "-D DIMY=<Value of threads_y>"
+///         }
+///         );
+/// \endcode
+///
+/// \param[in] kernelName is the name of the kernel qualified as kernel in code
+/// \param[in] sources is the list of source strings to be compiled if required
+/// \param[in] templateArgs is a vector of strings containing stringified names
+///            of the template arguments of OpenCL kernel to be compiled.
+/// \param[in] compileOpts is a vector of strings that enables the user to
+///            add definitions such as `-D<NAME>` or `-D<NAME>=<VALUE>` for
+///            the kernel compilation.
+///
+Kernel getKernel(const std::string& kernelName,
```

Comment:
There are couple of things being addressed here, not just unifying cache APIs of all backends. Firstly, cleanup of OpenCL cache API. The later part is trying to bring OpenCL cache API as close as possible to CUDA one so that they can be merged.

I am ware of Event class, and I have looked into trying to make the API similar. I had hit some roadblock w.r.t to some parameters/arguments. I am planning on looking into it again.